pkgs/applications/networking/browsers/mullvad-browser/default.nix

```diff
@@ -235,7 +264,16 @@ stdenv.mkDerivation rec {
     runHook postBuild
   '';
 
-  installPhase = ''
+  installPhase = if stdenv.isDarwin then ''
+    runHook preInstall
+
+    mkdir -p "$out/Applications/"
+    cp -r "Mullvad Browser/Mullvad Browser.app" "$out/Applications"
+    chmod +x "$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser"
+    makeWrapper "$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser" "$out/bin/mullvad-browser"
```

Comment:
On macOS, we should also ship `distribution/distribution.ini` & `distribution/policies.json`.

Reply:
You can verify that it's working in the Help -> About Mullvad Browser dialog where it'll mention NixOS.
I guess the "NixOS" mentions should be renamed to "Nixpkgs" or similar.

![image](https://github.com/NixOS/nixpkgs/assets/3314323/4de1a477-c4c7-47bc-bdc0-453ec32612e5)

Reply:
@felschr so have been digging into this and it seems Firefox has some helpful documentation:

https://support.mozilla.org/en-US/kb/deploying-firefox-customizations-macos

I added this to the installPhase, but was unable to get things working (no matter what, MacOS complained the package was broken):

```
    /usr/bin/xattr -r -d com.apple.quarantine "$out/Applications/Mullvad Browser.app"
    install -Dvm644 ${distributionIni} "$out/Applications/Mullvad Browser.app/Resources/distribution/distribution.ini"
    install -Dvm644 ${policiesJson} "$out/Applications/Mullvad Browser.app/Resources/distribution/policies.json"
```

Tbh, I don't see it being _hugely_ important to provide `distribution.ini` or `policies.json` while having to circumvent Apple's quarantining.

However, if you (or anyone) knows how to get around this, I'd be happy to try other solutions.

Reply:
The correct seems to be `/Applications/Mullvad Browser.app/Contents/Resources/distribution/distribution.ini`:
https://github.com/mozilla/dinobuildr/blob/ceb419c13daffa97cd6d7a234af23d9aaf7286a7/resources/firefox-bookmarks.sh#L19

Reply:
Right, in this case I'm attempting to modify the .app in $out instead of moving Mullvad Browser.app to the privileged, system-wide /Applications folder (which would require elevated privileges to edit and shouldn't be necessary just to execute the binary in the .app).

So for now, would it be reasonable to not include the distribution.ini and policies.json and just leave that section commented out?

I reached out on Matrix to see if anyone has other ideas/suggestions.

Reply:
Looks like someone came up with a potential solution that I will try tomorrow!

https://matrix.to/#/!lheuhImcToQZYTQTuI%3Anixos.org/%246Q-M8yqPp44v3zMYWmnlbJa89mbWivA4Lr58552Luf4


> If you’re repackaging an upstream app bundle, and it’s notarized, making any changes to it will flag it as broken on macOS 13+. (edited)

Reply:
> Right, in this case I'm attempting to modify the .app in $out instead of moving Mullvad Browser.app to the privileged, system-wide /Applications folder

What I actually meant is the missing `/Contents/` part in the paths in your code snippet.

Reply:
Oh, shoot! I didn't even notice, thank you.

Reply:
It seems that `distribution.ini` has no impact on the MacOS version of the app. I ended up testing on my Homebrew installed Mullvad Browser:

<img width="775" alt="image" src="https://github.com/NixOS/nixpkgs/assets/18178614/cca43f28-9345-4ac3-957a-9d047cbc0c85">

<img width="736" alt="image" src="https://github.com/NixOS/nixpkgs/assets/18178614/4e7cba09-f10e-455c-937a-41b44c7643aa">

(However, from that screenshot you can observe that `policies.json` is working)

Reply:
Great to hear the `policies.json` is working. The `distribution.ini` we can live without, I think.

Reply:
I can't figure out how to get the Nix-built .app to cooperate with MacOS' quarantining after adding `distribution.ini` and `policies.json`. I think we should give up on this piece for now as `distribution.ini` seems to not have an impact and I don't think that updating the Nix-built `.app` will do anything since it will live in the Nix store.

Reply:
Is there any alternative location that Firefox/Tor Browser/Mullvad Browser looks for `policies.json` (i.e. `~/.mullvad/mullvadbrowser/policies.json`)? I can't find any documentation on it, but if there was, we could probably make a MacOS wrapper that does something like this:

```
cp ${policiesJson} "~/Library/Application Support/MullvadBrowser/policies.json"
$out/bin/mullvad-browser $@
```

Reply:
I don't know if there is another way for this.

Reply:
So, to test, I reverted Mullvad Browser to 13.0.7 to see what happened with the update mechanism (and maybe figure out if it would work reasonably without `policies.json`). It attempted to automatically download and install the update and then just didn't do anything. Then after closing the app, it wouldn't re-launch. 

So, getting the .app to work with `policies.json` is an absolute priority before this PR can be merged because the auto-update mechanism breaks Mullvad Browser when built with Nix. :thumbsdown:

Reply:
Made progress! Seems that code signing is the issue. I am able to launch the modified `Mullvad Browser.app` after removing the code signature with this command:

```
/usr/bin/codesign -fs - "$out/Applications/Mullvad Browser.app"
```

However, the browser isn't functional at all. It launches, but the tabs are all crashed and I can't browse to any websites. I'm sure there's some way to get around this -- but, I'm very happy to find that the issue isn't the file attributes in the app. 😄 

Hopefully a more experienced Nix+MacOS user appears to help me figure out what else I am missing.

Reply:
I think that at this point, Mullvad Browser's current packaging for MacOS makes it essentially incompatible with Nix. I have filed an upstream issue with Mullvad Browser to (hopefully) address this: https://github.com/mullvad/mullvad-browser/issues/219#issue-2103969667

I will plan on waiting for a response there before continuing with this. I _really_ hope they give my feedback some serious consideration. 🤞🏻

Reply:
Found this just now: https://support.mozilla.org/en-US/kb/managing-policies-macos-desktops

Gonna give it a try 🤞🏻

Reply:
@felschr I discovered that on MacOS you can break autoupdating for Mullvad Browser by making the directory it uses for updates user immutable. The directory in question is `~/Library/Application\ Support/MullvadBrowser/UpdateInfo/`.

Potentially, I could write a wrapper script that does something like this:

```
echo 'Disable auto-updates'
chflags -R uchg "$HOME/Library/Application Support/MullvadBrowser/UpdateInfo/"
"$out/Applications/Mullvad Browser.app/Contents/MacOS/mullvadbrowser"

echo 'Disabling auto-updates'
chflags -R nouchg "$HOME/Library/Application Support/MullvadBrowser/UpdateInfo/"
```

When this directory is immutable, Mullvad Browser fails with an "unknown error" when checking for updates (but the functionality of the app is unaffected).

Would this be an unreasonable workaround to implement? It's kind of hack-ey so my feelings won't be hurt if it's a no-go. :smile: